File: examples/basic_usage.rs
//! Basic usage examples for the ELID library

use elid::*;

fn main() {
    println!("=== ELID String Similarity Examples ===\n");

    // Example 1: Levenshtein Distance
    println!("1. Levenshtein Distance:");
    let dist = levenshtein("kitten", "sitting");
    println!("   Distance between 'kitten' and 'sitting': {}", dist);

    let similarity = normalized_levenshtein("kitten", "sitting");
    println!("   Normalized similarity: {:.2}", similarity);
    println!();

    // Example 2: Jaro-Winkler Similarity (good for names)
    println!("2. Jaro-Winkler Similarity (best for names):");
    let names = vec![
        ("Martha", "Marhta"),
        ("John Smith", "Jon Smith"),
        ("DIXON", "DICKSON"),
    ];

    for (name1, name2) in names {
        let jaro_sim = jaro(name1, name2);
        let jw_sim = jaro_winkler(name1, name2);
        println!("   '{}' vs '{}':", name1, name2);
        println!("     Jaro: {:.3}, Jaro-Winkler: {:.3}", jaro_sim, jw_sim);
    }
    println!();

    // Example 3: Hamming Distance (equal-length strings)
    println!("3. Hamming Distance (for equal-length strings):");
    let dna_sequences = vec![
        ("ACGTACGT", "ACGTACCT"),
        ("AAAAAAAA", "AAAAAAAA"),
        ("11111111", "00000000"),
    ];

    for (seq1, seq2) in dna_sequences {
        if let Some(dist) = hamming(seq1, seq2) {
            println!("   '{}' vs '{}': {} differences", seq1, seq2, dist);
        }
    }
    println!();

    // Example 4: OSA Distance (with transpositions)
    println!("4. OSA Distance (handles transpositions):");
    let pairs = vec![
        ("ca", "ac"),
        ("abcd", "acbd"),
        ("hello", "hallo"),
    ];

    for (str1, str2) in pairs {
        let osa = osa_distance(str1, str2);
        let lev = levenshtein(str1, str2);
        println!("   '{}' vs '{}':", str1, str2);
        println!("     OSA: {}, Levenshtein: {}", osa, lev);
    }
    println!();

    // Example 5: Finding best match
    println!("5. Finding Best Match:");
    let candidates = vec![
        "apple",
        "application",
        "apply",
        "apricot",
        "banana",
    ];

    let query = "app";
    let (idx, score) = find_best_match(query, &candidates);
    println!("   Query: '{}'", query);
    println!("   Best match: '{}' (score: {:.3})", candidates[idx], score);
    println!();

    // Example 6: Finding all matches above threshold
    println!("6. Finding All Matches Above Threshold:");
    let matches = find_matches_above_threshold(query, &candidates, 0.5);
    println!("   Query: '{}', Threshold: 0.5", query);
    println!("   Matches:");
    for (idx, score) in matches {
        println!("     - '{}' (score: {:.3})", candidates[idx], score);
    }
    println!();

    // Example 7: Using SimilarityOpts
    println!("7. Using SimilarityOpts (case-insensitive, trimmed):");
    let opts = SimilarityOpts {
        case_sensitive: false,
        trim_whitespace: true,
        ..Default::default()
    };

    let pairs = vec![
        ("  HELLO  ", "hello"),
        ("World", "WORLD"),
        ("\tTest\n", "test"),
    ];

    for (str1, str2) in pairs {
        let dist = levenshtein_with_opts(str1, str2, &opts);
        println!("   '{}' vs '{}': distance = {}", str1, str2, dist);
    }
    println!();

    // Example 8: Unicode support
    println!("8. Unicode Support:");
    let unicode_pairs = vec![
        ("café", "cafe"),
        ("你好", "您好"),
        ("🎉🎊", "🎉"),
    ];

    for (str1, str2) in unicode_pairs {
        let dist = levenshtein(str1, str2);
        let sim = jaro(str1, str2);
        println!("   '{}' vs '{}':", str1, str2);
        println!("     Levenshtein: {}, Jaro: {:.3}", dist, sim);
    }
    println!();

    // Example 9: Real-world scenario - Product search
    println!("9. Real-world Example - Product Search:");
    let products = vec![
        "iPhone 14 Pro Max",
        "iPhone 14 Pro",
        "iPhone 14",
        "iPhone 13 Pro",
        "Samsung Galaxy S23",
        "Google Pixel 7",
    ];

    let searches = vec!["iphone 14 pro", "galaxy", "pixel"];

    for search in searches {
        // Find matches with normalized comparison (case-insensitive)
        let mut scored_products: Vec<_> = products
            .iter()
            .enumerate()
            .map(|(i, product)| {
                let lower_product = product.to_lowercase();
                let score = best_match(search, &lower_product);
                (i, score)
            })
            .collect();

        scored_products.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());

        println!("   Search: '{}'", search);
        println!("   Top 3 matches:");
        for (idx, score) in scored_products.iter().take(3) {
            println!("     - {} (score: {:.3})", products[*idx], score);
        }
        println!();
    }

    println!("=== End of Examples ===");
}

File: examples/python/basic_usage.py
"""
Basic usage examples for ELID Python bindings.

Run with: python basic_usage.py

First, install the package:
  maturin develop
"""

import elid

print("=== ELID Python Example ===\n")

# Example 1: Levenshtein Distance
print("1. Levenshtein Distance:")
distance = elid.levenshtein("kitten", "sitting")
print(f"   Distance between 'kitten' and 'sitting': {distance}")

similarity = elid.normalized_levenshtein("kitten", "sitting")
print(f"   Normalized similarity: {similarity:.2f}\n")

# Example 2: Jaro-Winkler Similarity
print("2. Jaro-Winkler Similarity (best for names):")
names = [
    ("Martha", "Marhta"),
    ("John Smith", "Jon Smith"),
    ("DIXON", "DICKSON"),
]

for name1, name2 in names:
    jaro_sim = elid.jaro(name1, name2)
    jw_sim = elid.jaro_winkler(name1, name2)
    print(f"   '{name1}' vs '{name2}':")
    print(f"     Jaro: {jaro_sim:.3f}, Jaro-Winkler: {jw_sim:.3f}")
print()

# Example 3: Finding best match
print("3. Finding Best Match:")
candidates = ["apple", "application", "apply", "apricot", "banana"]
query = "app"

result = elid.find_best_match(query, candidates)
print(f"   Query: '{query}'")
print(
    f"   Best match: '{candidates[result['index']]}' (score: {result['score']:.3f})\n"
)

# Example 4: Finding all matches above threshold
print("4. Finding All Matches Above Threshold:")
matches = elid.find_matches_above_threshold(query, candidates, 0.5)
print(f"   Query: '{query}', Threshold: 0.5")
print("   Matches:")
for match in matches:
    print(f"     - '{candidates[match['index']]}' (score: {match['score']:.3f})")
print()

# Example 5: Product Search Simulation
print("5. Product Search Simulation:")
products = [
    "iPhone 14 Pro Max",
    "iPhone 14 Pro",
    "iPhone 14",
    "iPhone 13 Pro",
    "Samsung Galaxy S23",
    "Google Pixel 7",
]

searches = ["iphone 14 pro", "galaxy", "pixel"]

for search in searches:
    # Score all products
    scored = [
        {
            "index": i,
            "product": product,
            "score": elid.best_match(search.lower(), product.lower()),
        }
        for i, product in enumerate(products)
    ]

    # Sort by score descending
    scored.sort(key=lambda x: x["score"], reverse=True)

    print(f"   Search: '{search}'")
    print("   Top 3 matches:")
    for match in scored[:3]:
        print(f"     - {match['product']} (score: {match['score']:.3f})")
    print()

# Example 6: Using SimilarityOpts
print("6. Using SimilarityOpts (case-insensitive, trimmed):")
opts = elid.SimilarityOpts(case_sensitive=False, trim_whitespace=True)

pairs = [
    ("  HELLO  ", "hello"),
    ("World", "WORLD"),
    ("\tTest\n", "test"),
]

for str1, str2 in pairs:
    dist = elid.levenshtein_with_opts(str1, str2, opts)
    print(f"   '{str1}' vs '{str2}': distance = {dist}")
print()

# Example 7: Hamming Distance
print("7. Hamming Distance (equal-length strings):")
hamming_pairs = [
    ("ACGTACGT", "ACGTACCT"),
    ("hello", "hallo"),
]

for seq1, seq2 in hamming_pairs:
    dist = elid.hamming(seq1, seq2)
    if dist is not None:
        print(f"   '{seq1}' vs '{seq2}': {dist} differences")
print()

# Example 8: OSA Distance (with transpositions)
print("8. OSA Distance (handles transpositions):")
osa_pairs = [
    ("ca", "ac"),
    ("abcd", "acbd"),
]

for str1, str2 in osa_pairs:
    osa = elid.osa_distance(str1, str2)
    lev = elid.levenshtein(str1, str2)
    print(f"   '{str1}' vs '{str2}':")
    print(f"     OSA: {osa}, Levenshtein: {lev}")
print()

# Example 9: Spell Checker
print("9. Simple Spell Checker:")
dictionary = ["receive", "believe", "achieve", "receive", "ceiling", "deceive"]
misspellings = ["recieve", "beleive", "achive", "cieling"]

for misspelling in misspellings:
    result = elid.find_best_match(misspelling, dictionary)
    suggestion = dictionary[result["index"]]
    print(f"   '{misspelling}' → '{suggestion}' (confidence: {result['score']:.2%})")
print()

# Example 10: Name Deduplication
print("10. Name Deduplication:")
names = [
    ("John Smith", "Jon Smith"),
    ("Robert Johnson", "Bob Johnson"),
    ("Mary Williams", "Marie Williams"),
    ("James Brown", "Jane Brown"),
]

threshold = 0.85
for name1, name2 in names:
    similarity = elid.jaro_winkler(name1, name2)
    is_duplicate = similarity > threshold
    status = "✓ Likely duplicate" if is_duplicate else "✗ Different people"
    print(f"   '{name1}' vs '{name2}': {similarity:.3f} - {status}")

print("\n=== End of Examples ===")
